lib/workflow/backend/backend.go

```diff
@@ -0,0 +1,137 @@
+package backend
+
+import (
+	"context"
+	"time"
+)
+
+type Driver interface {
+	EnqueueWorkflow(ctx context.Context, workflow *WorkflowInstance) error
+	DequeueWorkflow(ctx context.Context, workflowNames []string) (*WorkflowEvent, WorkflowCompleter, error)
+
+	DequeueTask(ctx context.Context, taskName string) (*Task, TaskCompleter, error)
+
+	CreateWorkflowSchedule(ctx context.Context, scheduleName string, workflowName string, parameters []byte, enabled bool, recurrence string, nextRunAt time.Time) error
+	GetDueRecurringWorkflow(ctx context.Context) (*Schedule, RecurringWorkflowCompleter, error)
+	GetNextScheduledWorkflow(ctx context.Context) (*Schedule, error)
+	UpdateWorkflowScheduleByID(ctx context.Context, id int64, opts WorkflowScheduleUpdateOpts) error
+	UpdateWorkflowScheduleByName(ctx context.Context, scheduleName string, workflowName string, opts WorkflowScheduleUpdateOpts) error
+
+	GetScheduledWorkflowParameters(ctx context.Context, scheduleName string, workflowName string) ([]byte, error)
+	GetScheduledWorkflowRecurrence(ctx context.Context, scheduleName string, workflowName string) (string, error)
+
+	ListWorkflowSchedules(ctx context.Context) ([]*Schedule, error)
+
+	Init() error
+	Close() error
+}
+
+type WorkflowInstanceStatus string
+
+const (
+	WorkflowInstanceStatusRunning   WorkflowInstanceStatus = "running"
+	WorkflowInstanceStatusAbandoned WorkflowInstanceStatus = "abandoned"
+)
+
+type WorkflowInstance struct {
+	InstanceName string
+	WorkflowName string
+	Status       WorkflowInstanceStatus
+	Parameters   []byte
+	Payload      []byte
+}
+
+type WorkflowEventType string
+
+const (
+	WorkflowStart  WorkflowEventType = "start"
+	TaskComplete   WorkflowEventType = "task_complete"
+	Cancel         WorkflowEventType = "cancel"
+	TasksAbandoned WorkflowEventType = "tasks_abandoned"
+)
+
+type TaskStatusType string
+
+const (
+	TaskStatusSuccess TaskStatusType = "success"
+	TaskStatusFailed  TaskStatusType = "failed"
+)
+
+type WorkflowEvent struct {
+	InstanceID         int64
```

Comment:
It might be cleaner for this to live on the implementation of WorkflowCompleter since that is the only thing that should depend on the internal state of the backend?